src/main.rs

```diff
@@ -34,9 +26,30 @@ async fn entry() -> Result<()> {
 
     info!("Initializing server...");
 
-    start_server().await?;
+    let server_handle = start_server().await?;
+
+    let need_to_kill = select! {
+        server_result = server_handle => {
+            match server_result.expect("join_error") {
+                Ok(_) => {
+                    info!("Server exited successfully!");
+                }
+                Err(e) => {
+                    error!("Server exited with an error");
+                    error!("{}", e);
+                }
+            }
+            false
+        },
+        _ = tokio::signal::ctrl_c() => {
+            info!("Received ctrl+c.. Shutting down..");
+            true
+        }
+    };
 
-    tokio::signal::ctrl_c().await?;
+    if need_to_kill {
+        kill_all_systems().await?;
+    }
```

Comment:
Why not just remove `need_to_kill` and put `kill_all_systems().await?` under `_ = tokio::signal::ctrl_c()` ?

Reply:
You cant have blocks of different types in select! and no two async blocks ever have the same type.